### ISSUE TITLE
Name of sorting functions

### ISSUE THREAD

**Issue by blefloch** [latex3/latex3#914]
- `\tl_sort:Nn` **changes** the order of items in the token list variable `#1` to be sorted according to the comparison given as some code `#2`.  Similar commands exist for clist and seq.
- `\tl_sort:nN` **expandably** sorts the items of the n-type token list `#1` according to the conditional `#2`, then leaves them in the input stream.

This is similar to how `\tl_reverse:N` changes `#1` while `\tl_reverse:n` is expandable and leaves the result in the input stream.  There may be other examples.  In retrospect, having a protected command and an expandable one share the same name, with only the signature changing, is probably a bad idea.  One mnemonic devise is that the first argument is N-type (hence can be assigned) or n-type (hence cannot possibly be assigned, which means the result must be put in the input stream).

This becomes important in #842, @Skillmon implements an expandable `\clist_sort:nN` (for which the mnemonic still works) **and** a `\clist_sort:NN` (for which the mnemonic doesn't work).  The latter is not equivalent to `\clist_sort:VN` because empty items have already been removed from N-type clists.

I think we should consider using different names for the protected and the expandable versions.  Perhaps like Python `sort` (an assignment) versus `sorted` (an expression used in further calculations), so `\tl_sorted:nN` leaving the result in the input stream.  Perhaps `\tl_sort_and_use:nN`?

**Comment by car222222** [comment]
I agree that different name bodies would be better!  

Aside:  There are other places where the name of such expandable commands do not clearly imply that 'this simply expands', leaving its result in the stream.  Maybe use of the past-participle of the verb would help there too?

**Comment by FrankMittelbach** [comment]
I agree that it was probably a bad idea. From the proposals I rather like `\<module>_<action>_and_use:...`.
 I think the "and use" is a clear indicator for being expandable and I think it fits well into existing names. Granted it is a bit longish, but I think it is better to have  `_<action>:..` and `_<action>_and_use:...`  than  `_<action>:..` and `_<action>ed:...`

**Comment by josephwright** [comment]
Wouldn't `\<thing>_use_sorted:N` work?

**Comment by josephwright** [comment]
I see @FrankMittelbach doesn't think so :)

**Comment by FrankMittelbach** [comment]
> Wouldn't `\<thing>_use_sorted:N` work?

that's a new suggestion (as far as I can tell), saves you one letter :-) would work for me too. But are you sure that works in all cases? (I thought we have some "actions" that are not proper verbs, no?)

**Comment by car222222** [comment]
`_use_<action>ed:` seems good, at least for some cases of <action>, such as 'sort'.

But I can also see the attraction of `_use:` being at the end.

`<action>_and_` does not feel right.

**Comment by car222222** [comment]
There is usually a verb around somewhere.

**Comment by josephwright** [comment]
OK, so we go with `use_sorted`?

**Comment by FrankMittelbach** [comment]
I don't mind either way as long as it is consistently used across the board.

**Comment by blefloch** [comment]
I don't like `use_sorted` very much, and would perhaps prefer `sort_use`, meaning "sort" then "use".  We have a few other `verb_verb` commands, like `\box_use_drop:N`, or `\tl_set_rescan:Nnn`, or `\tl_(trim_spaces)_apply:nN` which roughly speaking combine two operations.

We could also avoid the issue by only providing `\tl_sort_apply:nNN` which would sort `#1` using the comparison `#2`, then call `#3{sorted tl}`.  Then people can call it with `#3` being `\exp_not:n` to get the behaviour of `\tl_sort:nN`.

**Comment by josephwright** [comment]
Looking back, it looks like we want `\<thing>_sort_use:N` for the expandable versions. I will look at this for the `tl`  module, and if there are others, let me know.

**Comment by josephwright** [comment]
> Imho, `sort_use` sounds unnatural, might I suggest `use_sorted` if `sorted` alone is not strong enough a difference?

Did you see the discussion ;)

**Comment by blefloch** [comment]
I have no opinion on naming, but `\tl_reverse:n` should probably be renamed at the same time.

**Comment by FrankMittelbach** [comment]
as in `\tl_reverse_use:N` you mean? Guess so.

**Comment by blefloch** [comment]
Yes, except the signature should be `:n`.

**Comment by FrankMittelbach** [comment]
yep, finger trouble ... (usually it is wrong the other way around)
